src/groups/mqb/mqbauthn/mqbauthn_authenticationcontroller.cpp

```diff
@@ -0,0 +1,483 @@
+            errorDescription
+                << "Anonymous credential mechanism '" << normAnonMech
+                << "' does not match any configured authenticator";
+            return rc_NO_MATCHING_AUTHENTICATOR;  // RETURN
+        }
+    }
+
+    return rc_SUCCESS;
+}
+
+int AuthenticationController::ensureDefaultAuthenticator(
+    bsl::ostream& errorDescription)
+{
+    // If no authenticators are configured, use AnonAuthenticator for
+    // anonymous pass authentication
+    if (d_authenticators.empty()) {
+        BALL_LOG_INFO << "No authenticators configured, using "
+                         "AnonAuthenticator as default";
+
+        AnonAuthenticatorPluginFactory anonFactory;
+        AuthenticatorMp authenticator = anonFactory.create(d_allocator_p);
+
+        bmqu::MemOutStream errorStream(d_allocator_p);
+        if (int status = authenticator->start(errorStream)) {
+            errorDescription << "Failed to start AnonAuthenticator: "
+                             << errorStream.str();
+            BMQTSK_ALARMLOG_ALARM("#AUTHENTICATION")
+                << "Failed to start default anonymous authenticator '"
+                << authenticator->name() << "' [rc: " << status << ", error: '"
+                << errorStream.str() << "']" << BMQTSK_ALARMLOG_END;
+            return status;  // RETURN
+        }
+
+        const bsl::string normMech =
+            normalizeMechanism(authenticator->mechanism(), d_allocator_p);
+        d_authenticators.emplace(normMech,
+                                 bslmf::MovableRefUtil::move(authenticator));
+    }
+
+    return 0;
+}
+
+AuthenticationController::AuthenticationController(
+    mqbplug::PluginManager* pluginManager,
+    bslma::Allocator*       allocator)
+: d_builtInPluginLibrary(
+      bslma::ManagedPtrUtil::allocateManaged<mqbauthn::PluginLibrary>(
+          allocator))
+, d_pluginManager_p(pluginManager)
+, d_isStarted(false)
+, d_allocator_p(allocator)
+{
+}
+
+int AuthenticationController::start(bsl::ostream& errorDescription)
+{
+    enum RcEnum {
+        // Enum for the various RC error categories
+        rc_SUCCESS             = 0,
+        rc_ALREADY_STARTED     = -1,
+        rc_INVALID_CONFIG      = -2,
+        rc_INIT_AUTHENTICATORS = -3
+    };
+
+    if (d_isStarted) {
+        errorDescription << "start() can only be called once on this object";
+        return rc_ALREADY_STARTED;
+    }
+
+    BALL_LOG_INFO << "Starting AuthenticationController";
+
+    int rc = setAnonymousCredential(errorDescription);
+    if (rc != 0) {
+        return rc_INVALID_CONFIG;  // RETURN
+    }
+
+    rc = initializeAuthenticators(errorDescription);
+    if (rc != 0) {
+        return rc * 10 + rc_INIT_AUTHENTICATORS;
+    }
+
+    // Initialize Authenticators from plugins
+
+    d_isStarted = true;
+    return rc_SUCCESS;
+}
+
+void AuthenticationController::stop()
+{
+    if (!d_isStarted) {
+        return;  // RETURN
+    }
+
```

Comment:
```suggestion
    BALL_LOG_INFO << "Stopping AuthenticationController";

```
If we log starting, we might log stopping as well